File: leetcode/114_Flatten_Binary_Tree_to_Linked_List.py
from unittest import TestCase
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __eq__(self, obj) -> bool:
        return isinstance(obj, TreeNode) and self.val == obj.val and self.left == obj.left and self.right == obj.right

    def __repr__(self) -> str:
        return f'({self.val} {self.left} {self.right})'

class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        if root is not None:
            stack = [root]
            last = TreeNode(0, None, None)
            while stack:
                node = stack.pop()
                if node.right is not None:
                    stack.append(node.right)
                if node.left is not None:
                    stack.append(node.left)
                node.left, node.right = None, None
                last.right = node
                last = node

class FlattenTest(TestCase):
    def test_1(self):
        root = TreeNode(1, TreeNode(2, TreeNode(3), TreeNode(4)), TreeNode(5, right=TreeNode(6)))
        target = TreeNode(1, right=TreeNode(2, right=TreeNode(3, right=TreeNode(4, right=TreeNode(5, right=TreeNode(6))))))
        Solution().flatten(root)
        self.assertEqual(root, target)

    def test_2(self):
        root = None 
        Solution().flatten(root)
        self.assertEqual(root, None)

    def test_3(self):
        root = TreeNode(0)
        Solution().flatten(root)
        self.assertEqual(root, TreeNode(0))

File: leetcode/114_Flatten_Binary_Tree_to_Linked_List.rs
use std::rc::Rc;
use std::cell::RefCell;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None
        }
    }
}

struct Solution {}

impl Solution {
    pub fn flatten(root: &mut Option<Rc<RefCell<TreeNode>>>) {
        if let Some(node) = root {
            let mut stack = vec![node.clone()];
            let mut last = Rc::new(RefCell::new(TreeNode::new(0)));
            while let Some(node) = stack.pop() {
                let (left, right) = (node.borrow().left.clone(), node.borrow().right.clone());
                node.borrow_mut().left = None;
                node.borrow_mut().right = None;
                last.borrow_mut().right = Some(node.clone());
                last = node;
                if let Some(right) = right {
                    stack.push(right);
                }
                if let Some(left) = left {
                    stack.push(left);
                }
            }
        }
    }
}

mod flatten_test {
    use super::*;

    #[test]
    fn test_1() {
    }    
}